### PR TITLE
DOC: pd.read_csv doc-string clarification #11555

### LINKED ISSUE #11555: DOC: pd.read_csv doc-string clarification
need to update the docs for ``read_csv`` to match the doc-string. Let's make sure they have the same text / examples as much as possible (and are in the same order and such).

In particular, the ``header`` option doesn't have ``header='infer'`` which is the default in the docs, but it exists in the doc-string.

Further, should show what this does:

```
       if kwds.get('header', 'infer') == 'infer':
            kwds['header'] = 0 if kwds.get('names') is None else None
```

meaning if the the ``header`` kw is not specified, this it is set to the first line if no ``names`` are specified, else ``None``.

### PR DESCRIPTION
**Pull request by frankcleary**
closes #11555 
Updated IO Tools documentation for read_csv() and read_table() to be consistent with the doc-string.

### CODE REVIEW

**Comment by jreback** [comment]
what exactly changed here?

**Comment by jreback** [comment]
can you verify that these are in numpy-doc format, see [here](https://github.com/numpy/numpy/blob/master/doc/HOWTO_DOCUMENT.rst.txt)

mostly it comes down to a parameter like:

```
parm : datatype, optional
     description
```

**Comment by jreback** [comment]
```
Parameters

Description of the function arguments, keywords and their respective types.

Parameters
----------
x : type
    Description of parameter `x`.
Enclose variables in single backticks. The colon must be preceded by a space, or omitted if the type is absent.

For the parameter types, be as precise as possible. Below are a few examples of parameters and their types.

Parameters
----------
filename : str
copy : bool
dtype : data-type
iterable : iterable object
shape : int or tuple of int
files : list of str
If it is not necessary to specify a keyword argument, use optional:

x : int, optional
Optional keyword parameters have default values, which are displayed as part of the function signature. They can also be detailed in the description:

Description of parameter `x` (the default is -1, which implies summation
over all axes).
When a parameter can only assume one of a fixed set of values, those values can be listed in braces, with the default appearing first:

order : {'C', 'F', 'A'}
    Description of `order`.
When two or more input parameters have exactly the same type, shape and description, they can be combined:

x1, x2 : array_like
    Input arrays, description of `x1`, `x2`.
```

**Comment by frankcleary** [comment]
In terms of what was changed: I updated the IO Tools description of `read_csv()` that appears [here](http://pandas.pydata.org/pandas-docs/stable/io.html#csv-text-files) to match the doc string. I thought that's what the bug was getting at. I also slightly changed the wording describing the default behavior of the `header=` option.

I can go through and make any changes to put the doc string in numpy format, would you prefer that as a separate PR and/or should I drop the IO Tools doc changes?

**Comment by jreback** [comment]
@frankcleary thanks. can you enumerate what changed (here), though. just for my edification.

yes, would like to conform to numpy-doc in this PR.

**Comment by frankcleary** [comment]
Makes sense, I'll do that and update this PR later this week.

**Comment by jorisvandenbossche** [comment]
@jreback The list is in io.rst, not in the docstring. So I think it does not need to follow the numpy-doc format?

**Comment by jreback** [comment]
@jorisvandenbossche hmm, good point. I am not sure *what* we actually do as we don't have too many places where we go thru the args of functions. is there a consistent pattern now (anywhere)? any reason *not* to use a doc-string format?

**Comment by frankcleary** [comment]
The only other place I could find in the docs with a similar list of args is for `concat` in [merging.rst](http://pandas.pydata.org/pandas-docs/stable/merging.html#concatenating-objects). This also appears to have lagged behind the [`concat` docstring](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.concat.html) in terms of updates (for example there's no  `copy=...`). I wonder if it's better to remove these lists from the *.rst documentation in favor of links to the docs generated from the docstrings themselves?

**Comment by jreback** [comment]
@frankcleary the ideal thing would be to render the doc-strings in-line. as the out-of-syncness is annoying. Though if you simply want to fix it we can make that another issue.

**Comment by frankcleary** [comment]
Looks like sphinx can do that, updated. Here's a sample of what the result looks like in the io docs:


screen shot 2015-12-13 at 9 39 32 pm

**Comment by jreback** [comment]
@jorisvandenbossche how's this look to you

**Comment by jorisvandenbossche** [comment]
Hmm, personally, if it looks exactly the same as the generated API page, I would just link to it rather than including it. 
But I also understand that keeping a separate list of the parameters up to date is more work, but I think it is nicer for the tutorial docs.

**Comment by jreback** [comment]
so after seeing this, I think I agree with @jorisvandenbossche 

let's revert back to just an updated list (and hopefully we can keep this updated).

@frankcleary can you update.

**Comment by frankcleary** [comment]
Sure, I'll get to it this week.

File: doc/source/io.rst [hunk 1]
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
...
-  - ``tupleize_cols``: boolean, default False, if False, convert a list of tuples
-    to a multi-index of columns, otherwise, leave the column index as a list of
-    tuples
-  - ``float_precision`` : string, default None. Specifies which converter the C
-    engine should use for floating-point values. The options are None for the
-    ordinary converter, 'high' for the high-precision converter, and
-    'round_trip' for the round-trip converter.
+  - ``filepath_or_buffer``: str or file handle /
+    :class:`python:io.StringIO`. The string could be a URL. Valid URL schemes
+    include http, ftp, s3, and file. For file URLs, a host is expected. For
```
Comment (jreback) [inline doc/source/io.rst]: should we mention the LocalPath here as well? (its in the above)

File: doc/source/io.rst [hunk 2]
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
...
+    instance, a local file could be ``'file://localhost/path/to/table.csv'``.
+  - ``sep``: str, defaults to ``','`` for
+    :func:`~pandas.io.parsers.read_csv`, ``\t`` for
+    :func:`~pandas.io.parsers.read_table`. Delimiter to use. If sep is ``None``,
+    will try to automatically determine this. Regular expressions are accepted.
```
Comment (jreback) [inline doc/source/io.rst]: mention that regexs force the python engine

File: doc/source/io.rst [hunk 3]
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
...
+    ``QUOTE_MINIMAL`` (0), ``QUOTE_ALL`` (1), ``QUOTE_NONNUMERIC`` (2) or
+    ``QUOTE_NONE`` (3). Default (``None``) results in ``QUOTE_MINIMAL``
+    behavior.
+  - ``skipinitialspace``: boolean, default ``False``. Skip spaces after
+    delimiter.
+  - ``escapechar``: str (length 1), default ``None``. One-character string
+    used to escape delimiter when quoting is ``QUOTE_NONE``.
+  - ``dtype``: Type name or dict of column -> type, default ``None``. Data type
+    for data or columns. E.g. ``{'a': np.float64, 'b': np.int32}`` (unsupported
+    with ``engine='python'``).
```
Comment (jreback) [inline doc/source/io.rst]: menton ``str`` or ``object`` to preserve and not interpret for dtypes

File: doc/source/io.rst [hunk 4]
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
...
+    1 char. See :class:`python:csv.Dialect` documentation for more details.
+  - ``header``: int or list of ints, default ``'infer'``.  Row number(s) to use as
+    the column names, and the start of the data. Default behavior is as if
+    ``header=0`` if no ``names`` passed, otherwise as if ``header=None``.
+    Explicitly pass ``header=0`` to be able to replace existing names. The
+    header can be a list of integers that specify row locations for a
+    multi-index on the columns e.g. ``[0,1,3]``. Intervening rows that are not
+    specified will be skipped (e.g. 2 in this example is skipped). Note that
+    this parameter ignores commented lines and empty lines if
+    ``skip_blank_lines=True``, so header=0 denotes the first line of data
```
Comment (jreback) [inline doc/source/io.rst]: needs a ``-``

File: doc/source/io.rst [hunk 5]
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
...
+    file contains no header row, then you should explicitly pass
+    ``header=None``.
+  - ``prefix``: str, default ``None``. Prefix to add to column numbers when
+    no header, e.g. 'X' for X0, X1, ...
+  - ``na_values``: str, list-like or dict, default ``None``. Additional strings
+    to recognize as NA/NaN. If dict passed, specific per-column NA values.
+  - ``true_values``: list, default ``None``. Values to consider as ``True``.
+  - ``false_values``: list, default ``None``. Values to consider as ``False``.
+  - ``keep_default_na``: bool, default ``True``. If na_values are specified and
+    keep_default_na is ``False`` the default NaN values are overridden,
```
Comment (jreback) [inline doc/source/io.rst]: maybe add in the list of default NA values here? (or in ``na_values``)

File: doc/source/io.rst [hunk 6]
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
...
+    default ``False``. If ``True`` -> try parsing the index. If ``[1, 2, 3]`` ->
+    try parsing columns 1, 2, 3 each as a separate date column. If ``[[1, 3]]``
+    -> combine columns 1 and 3 and parse as a single date column. If
+    ``{'foo' : [1, 3]}`` -> parse columns 1, 3 as date and call result 'foo'.
+    A fast-path exists for iso8601-formatted dates.
+  - ``keep_date_col``: boolean, default ``False``. If ``True`` and parse_dates
+    specifies combining multiple columns then keep the original columns.
+  - ``date_parser``: function, default ``None``. Function to use for converting
+    a sequence of string columns to an array of datetime instances. The default
+    uses dateutil.parser.parser to do the conversion. Pandas will try to call
```
Comment (jreback) [inline doc/source/io.rst]: backticks around ``dateutil.parser.parser`` *maybe* put in a refernce to it

File: doc/source/io.rst [hunk 7]
```diff
@@ -79,116 +79,140 @@ for some advanced strategies
...
+  - ``verbose``: boolean, default ``False``. Indicate number of NA values placed
+    in non-numeric columns.
+  - ``delimiter``: str, default ``None``. Alternative argument name for sep.
+    Regular expressions are accepted.
+  - ``encoding``: str, default ``None``. Encoding to use for UTF when
+    reading/writing (e.g. ``'utf-8'``). `List of Python standard encodings
+    <https://docs.python.org/3/library/codecs.html#standard-encodings>`_.
+  - ``squeeze``: boolean, default ``False``. If the parsed data only contains
+    one column then return a Series.
+  - ``na_filter``: boolean, default ``True``. Detect missing value markers
```
Comment (jreback) [inline doc/source/io.rst]: let's put the NA options (``na_values``,``na_filter``,``keep_default_na``) near each other (change the code as well to match this ordering)

**Comment by jreback** [comment]
@frankcleary looks good. couple of points. 

I would like to have the options in the same order as ``io.rst`` sections. grouping if possible (e.g. the NA values together). I think this will be more logical.

All that said the more important options should be closer to the front if at all possible (I think that is how we had it). e.g. things like ``index_col``, ``header``, ``sep``. 

So ok to reorg a bit (see if you can keep the code in sync as well). everything is passed by kw so that should be fine. Add a note in the whatsnew to say that the ordering of args was changed (in API section) as well.

**Comment by frankcleary** [comment]
Makes sense, thanks. I will incorporate all those suggestions next weekend and post an update by Monday Jan. 18.

**Comment by frankcleary** [comment]
Here's the order I'm thinking of in terms of logically organizing the arguments, along with the general idea behind the grouping I put them in. Any thoughts?

##### Basic
filepath_or_buffer
sep
delimiter
##### Column and index names and location
header
names
index_col
usecols
squeeze
prefix
mangle_dupe_cols
##### General parsing details
dtype
engine
converters
true_values
false_values
skipinitialspace
skiprows
skipfooter
nrows
##### NA handling
na_values
keep_default_na
na_filter
verbose
skip_blank_lines
##### Date handling
parse_dates
infer_datetime_format
keep_date_col
date_parser
dayfirst
thousands
decimal
##### Iteration
iterator
chunksize
##### Quoting and file format
compression
lineterminator
quotechar
quoting
escapechar
comment
encoding
dialect
tupleize_cols
##### Errors
error_bad_lines
warn_bad_lines

**Comment by jreback** [comment]
@frankcleary a reorg like that would great!

(note you can prob use the ``^^^^`` to make these sub-sections in the docs (I think)

**Comment by frankcleary** [comment]
I've got an update almost ready with the reorg and other suggested changes, but need to fix merge conflicts in the parsers.py docstring so I won't be able to finish it tonight. I will get it done this weekend.

Is this about what you're thinking in regard to the sections:

screen shot 2016-02-03 at 10 53 41 pm

**Comment by jorisvandenbossche** [comment]
Thanks for this!

One consideration:
- I think we should try to make the actual option name stand out a bit more (not it is a bit lost in the dense bullet point). Maybe using something like http://docutils.sourceforge.net/docs/user/rst/quickref.html#definition-lists ? (just an idea, would have to try out to see how it looks)

**Comment by frankcleary** [comment]
Sorry, I was trying to push a clean history and it looks things are finished for this PR (can't reopen on a closed PR with a force pushed branch). I opened a new one here: https://github.com/pydata/pandas/pull/12256

File: doc/source/io.rst (hunk 1 above)
Comment (frankcleary) [inline doc/source/io.rst]: Fixed

File: doc/source/io.rst (hunk 2 above)
Comment (frankcleary) [inline doc/source/io.rst]: Fixed

File: doc/source/io.rst (hunk 3 above)
Comment (frankcleary) [inline doc/source/io.rst]: Fixed

File: doc/source/io.rst (hunk 4 above)
Comment (frankcleary) [inline doc/source/io.rst]: It's actually part of the paragraph for the header option

File: doc/source/io.rst (hunk 5 above)
Comment (frankcleary) [inline doc/source/io.rst]: Fixed by inserting the actual values, although I'm not sure if the added complexity is worth it.

File: doc/source/io.rst (hunk 6 above)
Comment (frankcleary) [inline doc/source/io.rst]: Fixed

File: doc/source/io.rst (hunk 7 above)
Comment (frankcleary) [inline doc/source/io.rst]: Did the major reorg as suggested

**Pull request closed without merging**